### PR TITLE
Add a way to track Observable execution

### PR DESCRIPTION
**Pull request by Odrakir**
## What's this?
This comes from a problem @pdrj was having the other day with Observables. For now is just a draft, and it's loosely based on this: https://github.com/ReactiveX/RxSwift/blob/master/RxExample/RxExample/Services/ActivityIndicator.swift

It adds a new class and an rx operators to make it easier to track the execution of an Observable.

Still needs some touches and some tests, but I would like to know what you think.

## :squid: Permissions
🚫

### CODE REVIEW

File: Toggl.Core/Extensions/ObservableExtensions.cs [hunk 1]
```diff
@@ -49,5 +51,61 @@ public static IObservable<ITransition> OnErrorReturnResult<T>(this IObservable<I
...
+            return _source;
+        }
+
+        public void Dispose()
+        {
+            _dispose.Dispose();
+        }
+    }
+
+    public class ActivityObserver : IDisposable
```
Comment (heytherewill) [inline Toggl.Core/Extensions/ObservableExtensions.cs]:
```suggestion
    public class ExecutionObserver : IDisposable
```

Activity is a very specific Android keyword and I'd rather avoid confusion

File: Toggl.Core/Extensions/ObservableExtensions.cs [hunk 2]
```diff
@@ -49,5 +51,61 @@ public static IObservable<ITransition> OnErrorReturnResult<T>(this IObservable<I
...
+    {
+        public IObservable<bool> running { get; }
```
Comment (heytherewill) [inline Toggl.Core/Extensions/ObservableExtensions.cs]:
```suggestion
        public IObservable<bool> Running { get; }
```

File: Toggl.Core/Extensions/ObservableExtensions.cs [hunk 3]
```diff
@@ -49,5 +51,61 @@ public static IObservable<ITransition> OnErrorReturnResult<T>(this IObservable<I
...
+            runningSubject.OnNext(false);
+        }
+
+        internal IObservable<T> trackActivityOfObservable<T>(IObservable<T> source)
+        {
+            return Observable.Using(
+                () =>
+                {
+                    runningSubject.OnNext(true);
+                    return new ActivityToken<T>(source, runningSubject);
```
Comment (heytherewill) [inline Toggl.Core/Extensions/ObservableExtensions.cs]: Is this equivalent to this:

```csharp
internal IObservable<T> trackActivityOfObservable<T>(IObservable<T> source)
    => Observable.Using(
        () => 
        {
            runningSubject.OnNext(true);
            return Disposable.Create(() => runningSubject.OnNext(false));
        },
        _ => source);
```

Or does the introduced type provide some other benefit?

File: Toggl.Core/Extensions/ObservableExtensions.cs [hunk 4]
```diff
@@ -49,5 +51,61 @@ public static IObservable<ITransition> OnErrorReturnResult<T>(this IObservable<I
...
+            running = runningSubject.AsObservable();
+            runningSubject.OnNext(true);
+        }
+
+        public void Dispose()
+        {
+            runningSubject.OnNext(false);
+        }
+
+        internal IObservable<T> trackActivityOfObservable<T>(IObservable<T> source)
```
Comment (heytherewill) [inline Toggl.Core/Extensions/ObservableExtensions.cs]:
```suggestion
        internal IObservable<T> TrackActivityOfObservable<T>(IObservable<T> source)
```

**Comment by heytherewill** [comment]
Is this still needed/being worked on?

**Comment by heytherewill** [comment]
pinging @Odrakir for maximum annoyance

**Comment by Odrakir** [comment]
There's no need for this right now, and didn't quite managed to make it work, so I'm closing it.

**Pull request closed without merging**